flowredux/src/commonMain/kotlin/com/freeletics/flowredux/dsl/State.kt

```diff
@@ -0,0 +1,37 @@
+package com.freeletics.flowredux.dsl
+
+/**
+ * Allows to create [ChangeState] objects to change the state as a result of DSL
+ * methods like [InStateBuilderBlock.on] or  [InStateBuilderBlock.onEnter].
+ *
+ * The [snapshot] property can be used to access the current state at the time the action
+ * or event happened. Note that it might be outdated by the time the code runs and should
+ * never be used within [mutate] or to create an object that is passed to [override]
+ */
+public class State<InputState : Any>(
```

Comment:
I can imagine that people call their State classes with `State` suffix already. i.e. `LoadingState`  or `FooState`.
Then we have `State<LoadingState>` which reads a bit strange.

Maybe we could call the class `Snapshot` instead of `State`, then it would become `Snapshot<LoadingState>` which sounds in my ears slightly better.


```kotlin
public class Snapshot<InputState : Any> (
   val state : InputState // or maybe name this property "current" instead of "state" or just "value"
) { 
  ...  
} 
```

With `Snapshot` as name then probably `.mutate()` or `.override()` don't read that nice anymore ...

Also as we have discussed on slack, maybe in a follow-up iteration we might want to get to a point where return type of "handler" is also `State`. That could read a bit strange with renaming it to `Snapshot`( but also not terrible bad ) :  `fun handleOnEnter( snapshot : Snapshot<MyState>) : Snapshot<MyState>`


What do you think?

Reply:
I took a cursory look across the freeletics code base and most of our state subclasses don't have state in the name (parent class FooState but then its subclasses are just Loading/Error/Content), but I still get your point.

Renaming it to something else is fine for me, but I'm not sure about calling it `Snapshot`. As you said the methods don't read as nicely and also if we think about replacing the `ChangeState` class I think returning a snapshot is weird. The snapshot we have is a state that was the current state when your dsl method is called but might be not the current state anymore while your method executes and especially not when the `mutate` lambda is invoked. What you return is a future change to the state, I think that would work with the `State` naming and maybe others but with `Snapshot` it's confusing.

Reply:
Then let's keep `State`, I also dont have a better name ...

Then I think keeping `override` as name makes sense. I would still introduce 2 overload for `override()` as discussed above and keep mutate for pure "mutation"

Reply:
Changed `mutate` to return `InputState` and added an overload of `override` here [a1ba193](https://github.com/freeletics/FlowRedux/pull/306/commits/a1ba1937ad1d66410195f56894786906a33d0d5f)